src/jrd/ExtEngineManager.cpp

```diff
@@ -774,20 +901,124 @@ ExtEngineManager::Function::~Function()
 }
 
 
-void ExtEngineManager::Function::execute(thread_db* tdbb, UCHAR* inMsg, UCHAR* outMsg) const
+// Execute an external function starting with an inactive request and ending with an active one.
+void ExtEngineManager::Function::execute(thread_db* tdbb, Request* request, jrd_tra* transaction,
+	unsigned inMsgLength, UCHAR* inMsg, unsigned outMsgLength, UCHAR* outMsg) const
 {
-	EngineAttachmentInfo* attInfo = extManager->getEngineAttachment(tdbb, engine.get());
-	const MetaString& userName = udf->invoker ? udf->invoker->getUserName() : "";
-	ContextManager<IExternalFunction> ctxManager(tdbb, attInfo, function,
-		(udf->getName().package.isEmpty() ?
-			CallerName(obj_udf, udf->getName().identifier, userName) :
-			CallerName(obj_package_header, udf->getName().package, userName)));
+	AutoSetRestore2<Request*, thread_db> autoSetRequest(
+		tdbb, &thread_db::getRequest, &thread_db::setRequest, request);
 
-	EngineCheckout cout(tdbb, FB_FUNCTION, checkoutType(attInfo->engine));
+	EXE_activate(tdbb, request, transaction);
 
-	FbLocalStatus status;
-	function->execute(&status, attInfo->context, inMsg, outMsg);
-	status.check();
+	fb_assert(inMsgLength == udf->getInputFormat()->fmt_length);
+	fb_assert(outMsgLength == udf->getOutputFormat()->fmt_length);
+
+	// Validate input parameters (internal message).
+	validateParameters(tdbb, inMsg, true);
+
+	// If there is a need for an external input message, copy the internal message to it and switch the message.
+	if (extInputImpureOffset.has_value())
+	{
+		const auto extInMsg = request->getImpure<UCHAR>(extInputImpureOffset.value());
+		copyMessage(tdbb, udf->getInputFormat(), inMsg, extInputFormat, extInMsg);
+		inMsg = extInMsg;
+		///inMsgLength = extInputFormat->fmt_length;
+	}
+
+	// Initialize outputs in the internal message.
+	{	// scope
+		fb_assert(udf->getOutputFormat()->fmt_desc.getCount() / 2 == udf->getOutputFields().getCount());
+
+		// Initialize everything to NULL (-1).
+		memset(outMsg, 0xFF, udf->getOutputFormat()->fmt_length);
+
+		for (const auto paramNumber : impl->outDefaults)
+		{
+			const auto param = udf->getOutputFields()[paramNumber];
+			const MetaNamePair namePair(param->prm_field_source, "");
+			FieldInfo fieldInfo;
+
+			dsc* defaultValue = nullptr;
+
+			if (request->getStatement()->mapFieldInfo.get(namePair, fieldInfo) &&
+				fieldInfo.defaultValue)
+			{
+				defaultValue = EVL_expr(tdbb, request, fieldInfo.defaultValue);
+
+				if (request->req_flags & req_null)
+					defaultValue = nullptr;
```

Comment:
IIRC, this condition/assignment is redundant, `EVL_expr()` is guaranteed to return `nullptr` in this case.